File: solutions/R/ch17b_ex1.R
# =============================================================================
# Chapter 17b, Exercise 1: Recover a known effect with IPW and g-computation
# Simulate a confounded cohort (true log-odds effect -0.8) and show that naive
# regression is biased while IPW (stabilised weights) and g-computation recover
# the truth and agree with each other.
# =============================================================================
#
# NOTE: The chapter demonstrates this with WeightIt / marginaleffects. Those
# packages are not assumed installed here, so we implement IPW and g-computation
# by hand with base R glm(). This is fully transparent and targets exactly the
# same estimands.

# --- Libraries (all loaded at the top; base R only) ---
# (no external packages required)

set.seed(42)

# --- Simulate a confounded cohort of 2000 patients -------------------------
# frail is a single binary confounder that raises BOTH the probability of
# treatment and the probability of the bad binary outcome (death).
n     <- 2000
frail <- rbinom(n, 1, 0.4)

# Frail patients are much more likely to be treated (confounding by indication)
p_treat <- plogis(-0.4 + 1.6 * frail)
treat   <- rbinom(n, 1, p_treat)

# TRUE causal effect of treatment on the outcome: conditional log-odds = -0.8
TRUTH   <- -0.8
p_death <- plogis(-0.7 + 1.0 * frail + TRUTH * treat)
death   <- rbinom(n, 1, p_death)

dat <- data.frame(frail, treat, death)

cat("Cohort: n =", n,
    "| treated =", sum(treat),
    "| deaths =", sum(death), "\n")
cat("Treatment rate by frailty:\n")
print(round(tapply(dat$treat, dat$frail, mean), 3))
cat("\n")

# --- (a) NAIVE logistic regression -----------------------------------------
# Regress the outcome on treatment ONLY, ignoring the confounder.
naive     <- glm(death ~ treat, data = dat, family = binomial)
naive_est <- coef(naive)["treat"]

cat("=== (a) Naive estimate ===\n")
cat(sprintf("Naive log-odds (treat):   %+.3f\n", naive_est))
cat(sprintf("Truth:                    %+.3f\n", TRUTH))
cat(sprintf("Bias (naive - truth):     %+.3f\n\n", naive_est - TRUTH))
# The naive estimate is biased toward zero (or positive): frail patients are
# both preferentially treated and more likely to die, so treatment looks less
# protective than it truly is.

# --- (b1) IPW with STABILISED weights --------------------------------------
# Step 1: propensity model P(A = 1 | L).
ps_model <- glm(treat ~ frail, data = dat, family = binomial)
ps       <- predict(ps_model, type = "response")   # P(A=1 | L)

# Step 2: stabilised weights  sw = P(A=a) / P(A=a | L)
p_marg <- mean(dat$treat)                           # marginal P(A=1)
sw <- ifelse(dat$treat == 1,
             p_marg       / ps,
             (1 - p_marg) / (1 - ps))
cat("=== (b) IPW (stabilised weights) ===\n")
cat(sprintf("Max stabilised weight: %.2f (well under 20 -> positivity OK)\n",
            max(sw)))

# Step 3: weighted outcome model (a marginal structural model).
# (quasibinomial avoids a harmless "non-integer #successes" warning under weights;
#  point estimates are identical to binomial)
msm    <- glm(death ~ treat, data = dat, family = quasibinomial, weights = sw)
ipw_lo <- coef(msm)["treat"]                        # marginal log-odds

# Marginal risk difference from the weighted model.
ipw_rd <- mean(predict(msm, transform(dat, treat = 1), type = "response")) -
          mean(predict(msm, transform(dat, treat = 0), type = "response"))
cat(sprintf("IPW marginal log-odds:    %+.3f\n", ipw_lo))
cat(sprintf("IPW risk difference:      %+.4f\n\n", ipw_rd))

# --- (b2) G-COMPUTATION (standardisation) ----------------------------------
# Step 1: fit ONE outcome model with treatment + confounder.
out_model <- glm(death ~ treat + frail, data = dat, family = binomial)

# Step 2: predict everyone under treat=1 and under treat=0.
p1 <- predict(out_model, transform(dat, treat = 1), type = "response")
p0 <- predict(out_model, transform(dat, treat = 0), type = "response")

# Step 3: average and contrast.
gcomp_rd <- mean(p1) - mean(p0)                          # marginal risk diff
gcomp_lo <- qlogis(mean(p1)) - qlogis(mean(p0))          # marginal log-odds
cat("=== (b) G-computation ===\n")
cat(sprintf("Conditional treat coef:   %+.3f (recovers the data-generating -0.8)\n",
            coef(out_model)["treat"]))
cat(sprintf("G-comp marginal log-odds: %+.3f\n", gcomp_lo))
cat(sprintf("G-comp risk difference:   %+.4f\n\n", gcomp_rd))

# --- (c) Comparison --------------------------------------------------------
cat("=== (c) Summary: log-odds scale ===\n")
cat(sprintf("  Truth (conditional):    %+.3f\n", TRUTH))
cat(sprintf("  Naive:                  %+.3f  (biased)\n", naive_est))
cat(sprintf("  IPW (marginal):         %+.3f\n", ipw_lo))
cat(sprintf("  G-computation (marg.):  %+.3f\n", gcomp_lo))
cat("\nBoth adjusted estimates land close to the truth and to each other,\n")
cat("while the naive estimate is clearly biased. (The marginal log-odds is\n")
cat("very slightly attenuated relative to the conditional -0.8 because the\n")
cat("logistic odds ratio is non-collapsible; the g-computation outcome-model\n")
cat("coefficient recovers the conditional -0.8 exactly.)\n")

File: solutions/python/ch17b_ex1.py
# =============================================================================
# Chapter 17b, Exercise 1: Recover a known effect with IPW and g-computation
# Simulate a confounded cohort (true log-odds effect -0.8) and show that naive
# regression is biased while IPW (stabilised weights) and g-computation recover
# the truth and agree with each other.
# =============================================================================
#
# NOTE: The chapter demonstrates this with WeightIt / marginaleffects (R) and
# statsmodels (Python). Here we implement IPW and g-computation by hand with
# statsmodels so every step is transparent and targets the same estimands.

import numpy as np
import pandas as pd
import statsmodels.api as sm
import statsmodels.formula.api as smf

np.random.seed(42)

# --- Simulate a confounded cohort of 2000 patients -------------------------
# frail is a single binary confounder that raises BOTH the probability of
# treatment and the probability of the bad binary outcome (death).
n = 2000
frail = np.random.binomial(1, 0.4, n)

# Frail patients are much more likely to be treated (confounding by indication)
p_treat = 1 / (1 + np.exp(-(-0.4 + 1.6 * frail)))
treat = np.random.binomial(1, p_treat)

# TRUE causal effect of treatment on the outcome: conditional log-odds = -0.8
TRUTH = -0.8
p_death = 1 / (1 + np.exp(-(-0.7 + 1.0 * frail + TRUTH * treat)))
death = np.random.binomial(1, p_death)

df = pd.DataFrame(dict(frail=frail, treat=treat, death=death))

print(f"Cohort: n = {n} | treated = {treat.sum()} | deaths = {death.sum()}")
print("Treatment rate by frailty:")
print(df.groupby("frail")["treat"].mean().round(3).to_string())
print()

def logit(p):
    return np.log(p / (1 - p))

# --- (a) NAIVE logistic regression -----------------------------------------
# Regress the outcome on treatment ONLY, ignoring the confounder.
naive = smf.logit("death ~ treat", data=df).fit(disp=0)
naive_est = naive.params["treat"]

print("=== (a) Naive estimate ===")
print(f"Naive log-odds (treat):   {naive_est:+.3f}")
print(f"Truth:                    {TRUTH:+.3f}")
print(f"Bias (naive - truth):     {naive_est - TRUTH:+.3f}\n")
# The naive estimate is biased toward zero (or positive): frail patients are
# both preferentially treated and more likely to die, so treatment looks less
# protective than it truly is.

# --- (b1) IPW with STABILISED weights --------------------------------------
# Step 1: propensity model P(A = 1 | L).
ps = smf.logit("treat ~ frail", data=df).fit(disp=0).predict(df)  # P(A=1 | L)

# Step 2: stabilised weights  sw = P(A=a) / P(A=a | L)
p_marg = df["treat"].mean()                                        # marginal P(A=1)
sw = np.where(df["treat"] == 1, p_marg / ps, (1 - p_marg) / (1 - ps))

print("=== (b) IPW (stabilised weights) ===")
print(f"Max stabilised weight: {sw.max():.2f} (well under 20 -> positivity OK)")

# Step 3: weighted outcome model (a marginal structural model).
msm = smf.glm("death ~ treat", data=df, family=sm.families.Binomial(),
              var_weights=sw).fit()
ipw_lo = msm.params["treat"]                                       # marginal log-odds

# Marginal risk difference from the weighted model.
ipw_rd = (msm.predict(df.assign(treat=1)).mean()
          - msm.predict(df.assign(treat=0)).mean())
print(f"IPW marginal log-odds:    {ipw_lo:+.3f}")
print(f"IPW risk difference:      {ipw_rd:+.4f}\n")

# --- (b2) G-COMPUTATION (standardisation) ----------------------------------
# Step 1: fit ONE outcome model with treatment + confounder.
out_model = smf.glm("death ~ treat + frail", data=df,
                    family=sm.families.Binomial()).fit()

# Step 2: predict everyone under treat=1 and under treat=0.
p1 = out_model.predict(df.assign(treat=1))
p0 = out_model.predict(df.assign(treat=0))

# Step 3: average and contrast.
gcomp_rd = p1.mean() - p0.mean()                                   # marginal risk diff
gcomp_lo = logit(p1.mean()) - logit(p0.mean())                     # marginal log-odds
print("=== (b) G-computation ===")
print(f"Conditional treat coef:   {out_model.params['treat']:+.3f} "
      "(recovers the data-generating -0.8)")
print(f"G-comp marginal log-odds: {gcomp_lo:+.3f}")
print(f"G-comp risk difference:   {gcomp_rd:+.4f}\n")

# --- (c) Comparison --------------------------------------------------------
print("=== (c) Summary: log-odds scale ===")
print(f"  Truth (conditional):    {TRUTH:+.3f}")
print(f"  Naive:                  {naive_est:+.3f}  (biased)")
print(f"  IPW (marginal):         {ipw_lo:+.3f}")
print(f"  G-computation (marg.):  {gcomp_lo:+.3f}")
print()
print("Both adjusted estimates land close to the truth and to each other,")
print("while the naive estimate is clearly biased. (The marginal log-odds is")
print("very slightly attenuated relative to the conditional -0.8 because the")
print("logistic odds ratio is non-collapsible; the g-computation outcome-model")
print("coefficient recovers the conditional -0.8 exactly.)")
